modin/core/execution/utils.py

```diff
@@ -29,3 +29,39 @@ def set_env(**environ):
     finally:
         os.environ.clear()
         os.environ.update(old_environ)
+
+
+# Check if the function already exists to avoid circular imports
+if "remote_function" not in dir():
+    from modin.config import Engine
+
+    if Engine.get() == "Ray":
+        from modin.core.execution.ray.common import RayWrapper
+
+        _remote_function_wrapper = RayWrapper.put
+    elif Engine.get() == "Unidist":
+        from modin.core.execution.unidist.common import UnidistWrapper
+
+        _remote_function_wrapper = UnidistWrapper.put
+    elif Engine.get() == "Dask":
+        from modin.core.execution.dask.common import DaskWrapper
+
+        # The function cache is not supported for Dask
+        def remote_function(func):
+            return DaskWrapper.put(func)
+
+    else:
+
+        def remote_function(func):
+            return func
+
+    if "remote_function" not in dir():
+        _remote_function_cache = {}
+
+        def remote_function(func):  # noqa: F811
+            func_id = id(func.__code__)
```

Comment:
if a cached function uses something from an outer scope then it can lead to some problems, is there a way to detect this and not cache in this case? Or we can simply put a notice that one should not use the caching with such functions
```python

_cache = {}

def run_func(x):

    def fn_to_run():
        print(x)

    fn_id = id(fn_to_run.__code__)
    if fn_id in _cache:
        _cache[fn_id]()
    else:
        _cache[fn_id] = fn_to_run
        fn_to_run()

run_func("first func")
run_func("second func")

# Output:
# first func
# first func
```

Reply:
I think you can check for [`fn.__closure__`](https://www.geeksforgeeks.org/__closure__-magic-function-in-python/) to detect whether it uses something from an outer scope